---
title: "Analyse Orthofinder output"
output:
  html_document:
    df_print: paged
---


```{r}
library(ggtree)
library(tidyverse)
library(ggstance) # for horizontal versions of geoms
library(gplots)
library('RColorBrewer')
#devtools::install_github("jakobbossek/ggheatmap")
library(ggheatmap)
```

```{r}
#tree<-read.tree("/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep18/Species_Tree/SpeciesTree_rooted.txt")
#manually rerooted tree
tree<-read.tree("/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Cobs.phylogeny.rerooted2.tre")
tree$node.label[-1]<-round(as.numeric(tree$node.label[-1]),2)
tree$tip.label<-gsub(".longestIsoform","",tree$tip.label)
plot(tree)
ape::nodelabels()
tum<-phytools::force.ultrametric(tree)
ape::write.tree(tum,file="/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Cobs.phylogeny.rerooted.ultrametric.tre")
p<-ggtree(tree) + 
  geom_tiplab(align=T)+
 xlim_expand(c(0,.28), panel = "Tree") + theme_tree2()

```


```{r}
#heat<-read.csv("/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep18/Comparative_Genomics_Statistics/OrthologuesStats_one-to-one.tsv",sep="\t")
heat<-read.csv("/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep21/Comparative_Genomics_Statistics/OrthologuesStats_one-to-one.tsv",sep="\t")
rownames(heat)<-gsub(".longestIsoform","",heat$X)
colnames(heat)<-gsub(".longestIsoform","",colnames(heat))
hmcol<-colorRampPalette(brewer.pal(9, "RdBu"))(100)
hm<-heatmap.2(as.matrix(heat[-1]),trace = "none",col = hmcol,key = F)
cor.mat<-cor(heat[-1])
tmp<-subset(p$data,isTip==T)
dh<-as.matrix(heat[-1])
diag(dh)=NA
#dh[upper.tri(dh)]<-NA
```

```{r}


pN <- ggtree(tree) + geom_tiplab(size=5, align=TRUE, linesize=.5) + theme_tree2()
pp <- (pN + scale_y_continuous(expand=c(0, 0.18))) %>%
    gheatmap(dh[,tmp$label[order(tmp$y)]], offset=.02, width=1.1, colnames=T,low="steelblue",high="red",legend_title = "one-to-one orthologs")%>%
        scale_x_ggtree()

```


```{r}

#sums<-read.csv("/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep18/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv",sep="\t",nrows = 10)
sums<-read.csv("/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep21/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv",sep="\t",nrows = 10)
rownames(sums)<-sums$X
sums2<-as.data.frame(t(sums[,-1]))
sums2$id<-gsub(".longestIsoform","",rownames(sums2))
sumStack<-reshape2::melt(sums2,id.vars="id")


```

```{r}
# create some random tree and trait data

sumStack
# tree plot with barplot facet


p2<-facet_plot(p,
           panel = "'orphan genes'",
           data = subset(sumStack,variable=="Number of genes in species-specific orthogroups"),
           geom = geom_barh,
           mapping = aes(x = value, fill = as.factor(variable)),
           stat = "identity")+
  theme(legend.position = "none",)


p3<-facet_plot(p2,
           panel = "# of genes",
           data = subset(sumStack,variable=="Number of genes"),
           geom = geom_barh,
           mapping = aes(x = value, fill = as.factor(variable)),
           stat = "identity")


```
```{r}

#tr<-tree
#d3 <- data.frame(id = rep(tr$tip.label, each=2),
#                    value = abs(rnorm(length(rep(tr$tip.label, each=2)), mean=100, sd=50)),
#                    category = rep(LETTERS[1:2], length(rep(tr$tip.label, each=2))))
#p3 <- facet_plot(p, panel = 'Stacked Barplot', data = d3, 
#                geom = geom_barh, 
#                mapping = aes(x = value, fill = as.factor(category)), 
#                stat='identity' ) 
#p3
```
```{r}
p3
#dev.print(pdf,"/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep18/orthofinder1.pdf",width=8,height=5)
dev.print(pdf,"/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep21/orthofinder1.pdf",width=8,height=5)
```
```{r}
pp
#dev.print(pdf,"/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep18/orthofinder2.pdf",width=12,height=5)
dev.print(pdf,"/Users/lukas/sciebo/Projects/CardiocondylaGenome/CardioMinION/analyses/comparativeGenomics/Results_Sep21/orthofinder2.pdf",width=12,height=5)
```

